frontends/pytorch/csrc/builder/graph_importer.h

```diff
@@ -0,0 +1,117 @@
+#ifndef NPCOMP_FRONTENDS_PYTORCH_CSRC_GRAPH_IMPORTER_H
+#define NPCOMP_FRONTENDS_PYTORCH_CSRC_GRAPH_IMPORTER_H
+
+#include <memory>
+
+#include "../pybind.h"
+#include "func_builder.h"
+
+#include "mlir-c/IR.h"
+
+#include <torch/csrc/jit/api/compilation_unit.h>
+#include <torch/csrc/jit/ir/ir.h>
+
+namespace torch_mlir {
+
+/// Main entry-point for importing graphs.
+/// In torch terminology, a Graph is a function. We materialize it as at
+/// least one MLIR function, but may specialize multiple versions of it,
+/// depending on usage.
+///
+/// Since graph functions typically have enough annotations for the most
+/// generic form of every type (i.e. Tensor, List, etc), and since we often
+/// want to multi-version specific specializations, we take the approach of
+/// generating a '$generic' suffixed function at that level and then generate
+/// the actual named function with using a 'numpy.generic_call' op to invoke
+/// the generic function with metadata controlling how it is legal to
+/// specialize. This leaves the process of inlining and expanding the
+/// specializations to compiler passes.
+class GraphImporter : public std::enable_shared_from_this<GraphImporter> {
+public:
+  struct MlirMappingOptions {
+    MlirContext context;
+    std::string genericFuncName;
+    std::string funcName;
+    TypeMapper &typeMapper;
+    FuncBuilder::Inserter &inserter;
+  };
+  /// Construct an importer.
+  GraphImporter(std::shared_ptr<torch::jit::Graph> graph,
+                MlirMappingOptions mappingOptions);
+
+  /// Helper to create a graph importer from a traced/scripted python function.
+  /// If the funcName of the mapping options is not set, it is set from the
+  /// function name. It is the responsibility of the caller to ensure that the
+  /// funcObj and associated graph outlives this instance.
+  static std::shared_ptr<GraphImporter>
+  forPythonJitFunc(torch::jit::Function *function,
+                   MlirMappingOptions mappingOptions);
+
+  /// Initialize for import. This is separate from the constructor purely for
+  /// ergonomics and must be called post-construction. Initialization activities
+  /// that throw go here.
+  void initialize();
+
+  /// Imports the generic function into the module.
+  void importGenericFunc();
+
+private:
+  // A scope of Graph Value * to corresponding MlirValue. Scopes nest
+  // region-wise. Note that in PyTorch, the thing called 'Block' is analagous
+  // to a capturing MLIR region.
+  struct Scope {
+    Scope() = default;
+    Scope(Scope *prev) : prev(prev) {}
+    llvm::DenseMap<torch::jit::Value *, MlirValue> valueMap;
+    Scope *prev = nullptr;
+
+    void bindValue(torch::jit::Value *torchValue, MlirValue value);
+    MlirValue findValue(torch::jit::Value *torchValue);
+    MlirValue findRequiredValue(MlirLocation loc,
+                                torch::jit::Value *torchValue);
+  };
+
+  struct Importer {
+    Importer(GraphImporter &parent, FuncBuilder *funcBuilder, MlirBlock block,
```

Comment:
I somehow feel that `node` should be the first argument/member, as it's the main thing that is being acted on -- we are importing `node`. The other stuff is just metadata/utilities for how to do that.